{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Genomic Assembly & Variant Calling of Chloroplasts from 12 Tomato Cultivars </font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Assembly of Sequence Reads Against Reference Genome </font>\n",
    "<img src=\"SequenceMapping.jpg\" /> <br> Source:  Metagenomics <br> http://www.metagenomics.wiki/pdf/definition/coverage-read-depth"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Notional Alignment Example </font>\n",
    "<img src=\"AlignmentExample.jpg\" /> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Bioinformatics Tools Pipeline </font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"PipelineFlowchart2.jpg\"  style=\"width:90%;height:90%\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Chloroplast Sequence Reads from 12 Tomato Cultivars </font> \n",
    "<img src=\"cultivars.jpg\"  style=\"width:1000px;height:350\"/> <br> Source:  National Center for Biotechnology (NCBI) Sequence Read Archive <br>  https://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP052069&o=acc_s%3Aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Analysis of Sequence Reads (Sequence Quality) </font> \n",
    "<img src=\"PerBaseSequenceQuality.jpg\" /> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Analysis of Sequence Reads (Sequence Length Distribution) </font> \n",
    "<img src=\"SequenceLengthDistribution.jpg\" /> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Reference Genome (NC_007898) </font> \n",
    "<img src=\"ReferenceGenome.jpg\" /> <br> Source:  NCBI Genome <br> https://www.ncbi.nlm.nih.gov/genome/?term=NC_007898 <br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Jupyter Notebook Extension for Integrative Genomics Viewer (IGV) </font>\n",
    "##  Enables viewing/comparison of genomic variation across multiple subjects\n",
    "<img src=\"igvExtension.jpg\" /> <br> Source:  GitHub - igvteam/igv-jupyter <br> https://github.com/igvteam/igv-jupyter#igv-jupyter-extension"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Import Python Modules & Retrieve Primary Directory Path </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import igv\n",
    "sequenceDirectoryFilePath = os.path.join(os.getcwd(), \"Pipeline\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Download Genomic Sequence Read Files from NCBI SRA </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating new directory at ./Pipeline for sequence reads.\n",
      "Read 120678 spots for SRR1763770\n",
      "Written 120678 spots for SRR1763770\n",
      "Read 153385 spots for SRR1763771\n",
      "Written 153385 spots for SRR1763771\n",
      "Read 69657 spots for SRR1763772\n",
      "Written 69657 spots for SRR1763772\n",
      "Read 213112 spots for SRR1763773\n",
      "Written 213112 spots for SRR1763773\n",
      "Read 61625 spots for SRR1763774\n",
      "Written 61625 spots for SRR1763774\n",
      "Read 150218 spots for SRR1763775\n",
      "Written 150218 spots for SRR1763775\n",
      "Read 71393 spots for SRR1763776\n",
      "Written 71393 spots for SRR1763776\n",
      "Read 143622 spots for SRR1763777\n",
      "Written 143622 spots for SRR1763777\n",
      "Read 71952 spots for SRR1763778\n",
      "Written 71952 spots for SRR1763778\n",
      "Read 55330 spots for SRR1763779\n",
      "Written 55330 spots for SRR1763779\n",
      "Read 155683 spots for SRR1763780\n",
      "Written 155683 spots for SRR1763780\n",
      "Read 68825 spots for SRR1763781\n",
      "Written 68825 spots for SRR1763781\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "chmod +x ./getChloroplastReads.sh\n",
    "./getChloroplastReads.sh ./Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> List Downloaded Read Files </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BC_17.fastq  BC_20.fastq  BC_22.fastq  BC_24.fastq  BC_26.fastq  BC_29.fastq\r\n",
      "BC_19.fastq  BC_21.fastq  BC_23.fastq  BC_25.fastq  BC_28.fastq  BC_30.fastq\r\n"
     ]
    }
   ],
   "source": [
    "ls ./Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Clean & Trim Read Files </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing fastq files within directory ./Pipeline/.\n",
      "File -> BC_17.fastq\n",
      "Working on -> BC_17\n",
      "File -> BC_19.fastq\n",
      "Working on -> BC_19\n",
      "File -> BC_20.fastq\n",
      "Working on -> BC_20\n",
      "File -> BC_21.fastq\n",
      "Working on -> BC_21\n",
      "File -> BC_22.fastq\n",
      "Working on -> BC_22\n",
      "File -> BC_23.fastq\n",
      "Working on -> BC_23\n",
      "File -> BC_24.fastq\n",
      "Working on -> BC_24\n",
      "File -> BC_25.fastq\n",
      "Working on -> BC_25\n",
      "File -> BC_26.fastq\n",
      "Working on -> BC_26\n",
      "File -> BC_28.fastq\n",
      "Working on -> BC_28\n",
      "File -> BC_29.fastq\n",
      "Working on -> BC_29\n",
      "File -> BC_30.fastq\n",
      "Working on -> BC_30\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TrimmomaticSE: Started with arguments:\n",
      " BC_17.fastq BC_17.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 150218 Surviving: 112654 (74.99%) Dropped: 37564 (25.01%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_19.fastq BC_19.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 55330 Surviving: 42038 (75.98%) Dropped: 13292 (24.02%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_20.fastq BC_20.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 153385 Surviving: 118838 (77.48%) Dropped: 34547 (22.52%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_21.fastq BC_21.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 213112 Surviving: 173352 (81.34%) Dropped: 39760 (18.66%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_22.fastq BC_22.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 71952 Surviving: 56073 (77.93%) Dropped: 15879 (22.07%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_23.fastq BC_23.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 143622 Surviving: 107867 (75.10%) Dropped: 35755 (24.90%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_24.fastq BC_24.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 61625 Surviving: 46847 (76.02%) Dropped: 14778 (23.98%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_25.fastq BC_25.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 69657 Surviving: 51760 (74.31%) Dropped: 17897 (25.69%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_26.fastq BC_26.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 68825 Surviving: 50448 (73.30%) Dropped: 18377 (26.70%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_28.fastq BC_28.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 71393 Surviving: 53566 (75.03%) Dropped: 17827 (24.97%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_29.fastq BC_29.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 120678 Surviving: 89684 (74.32%) Dropped: 30994 (25.68%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " BC_30.fastq BC_30.trim.fastq ILLUMINACLIP:/home/youngjb/Final_Project/ION_Adaptors.fa:4:10:3 SLIDINGWINDOW:30:20 MINLEN:25\n",
      "Automatically using 4 threads\n",
      "Using Medium Clipping Sequence: 'CCTCTCTATGGGCAGTCGGTGAT'\n",
      "Using Long Clipping Sequence: 'CCATATCATCCCTGCGTGTCTCCCACTCAG'\n",
      "ILLUMINACLIP: Using 0 prefix pairs, 2 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 155683 Surviving: 121582 (78.10%) Dropped: 34101 (21.90%)\n",
      "TrimmomaticSE: Completed successfully\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "chmod +x ./trimReads.sh\n",
    "./trimReads.sh ./Pipeline/ /home/youngjb/Final_Project/ION_Adaptors.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  <font color='blue'> List Cleaned/Trimmed Read Files </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BC_17.trim.fastq  BC_21.trim.fastq  BC_24.trim.fastq  BC_28.trim.fastq\r\n",
      "BC_19.trim.fastq  BC_22.trim.fastq  BC_25.trim.fastq  BC_29.trim.fastq\r\n",
      "BC_20.trim.fastq  BC_23.trim.fastq  BC_26.trim.fastq  BC_30.trim.fastq\r\n"
     ]
    }
   ],
   "source": [
    "ls ./Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'> Perform Remaining Pipeline Steps </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting process to generate .vcf associated files within directory /home/youngjb/Final_Project/Pipeline.\n",
      "Settings:\n",
      "  Output files: \"NC_007898.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  NC_007898.fasta\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 38865\n",
      "Using parameters --bmax 29149 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 29149 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 155461 (target: 29148)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 155461 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 155462 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 47640\n",
      "fchr[G]: 77564\n",
      "fchr[T]: 106500\n",
      "fchr[$]: 155461\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4246367 bytes to primary EBWT file: NC_007898.1.bt2\n",
      "Wrote 38872 bytes to secondary EBWT file: NC_007898.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 155461\n",
      "    bwtLen: 155462\n",
      "    sz: 38866\n",
      "    bwtSz: 38866\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 9717\n",
      "    offsSz: 38868\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 810\n",
      "    numLines: 810\n",
      "    ebwtTotLen: 51840\n",
      "    ebwtTotSz: 51840\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 38865\n",
      "Using parameters --bmax 29149 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 29149 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 155461 (target: 29148)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 155461 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 155462 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 47640\n",
      "fchr[G]: 77564\n",
      "fchr[T]: 106500\n",
      "fchr[$]: 155461\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4246367 bytes to primary EBWT file: NC_007898.rev.1.bt2\n",
      "Wrote 38872 bytes to secondary EBWT file: NC_007898.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 155461\n",
      "    bwtLen: 155462\n",
      "    sz: 38866\n",
      "    bwtSz: 38866\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 9717\n",
      "    offsSz: 38868\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 810\n",
      "    numLines: 810\n",
      "    ebwtTotLen: 51840\n",
      "    ebwtTotSz: 51840\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:01\n",
      "File -> BC_17.trim.fastq\n",
      "Working on -> BC_17.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:28:54,376]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_19.trim.fastq\n",
      "Working on -> BC_19.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:29:05,108]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_20.trim.fastq\n",
      "Working on -> BC_20.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:30:59,213]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_21.trim.fastq\n",
      "Working on -> BC_21.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:31:35,622]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_22.trim.fastq\n",
      "Working on -> BC_22.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:32:05,580]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_23.trim.fastq\n",
      "Working on -> BC_23.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:32:43,247]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_24.trim.fastq\n",
      "Working on -> BC_24.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:33:10,394]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_25.trim.fastq\n",
      "Working on -> BC_25.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:33:45,779]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_26.trim.fastq\n",
      "Working on -> BC_26.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:34:19,036]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_28.trim.fastq\n",
      "Working on -> BC_28.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:34:32,095]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_29.trim.fastq\n",
      "Working on -> BC_29.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:35:40,588]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n",
      "File -> BC_30.trim.fastq\n",
      "Working on -> BC_30.trim\n",
      "Using system JDK.\n",
      "WARN [2020-04-21T00:37:20,543]  [Globals.java:139] [main]  Development mode is enabled\n",
      "Done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building a SMALL index\n",
      "112654 reads; of these:\n",
      "  112654 (100.00%) were unpaired; of these:\n",
      "    16819 (14.93%) aligned 0 times\n",
      "    61970 (55.01%) aligned exactly 1 time\n",
      "    33865 (30.06%) aligned >1 times\n",
      "85.07% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "42038 reads; of these:\n",
      "  42038 (100.00%) were unpaired; of these:\n",
      "    29784 (70.85%) aligned 0 times\n",
      "    8665 (20.61%) aligned exactly 1 time\n",
      "    3589 (8.54%) aligned >1 times\n",
      "29.15% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "118838 reads; of these:\n",
      "  118838 (100.00%) were unpaired; of these:\n",
      "    11364 (9.56%) aligned 0 times\n",
      "    66870 (56.27%) aligned exactly 1 time\n",
      "    40604 (34.17%) aligned >1 times\n",
      "90.44% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "173352 reads; of these:\n",
      "  173352 (100.00%) were unpaired; of these:\n",
      "    122694 (70.78%) aligned 0 times\n",
      "    34150 (19.70%) aligned exactly 1 time\n",
      "    16508 (9.52%) aligned >1 times\n",
      "29.22% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "56073 reads; of these:\n",
      "  56073 (100.00%) were unpaired; of these:\n",
      "    13493 (24.06%) aligned 0 times\n",
      "    32277 (57.56%) aligned exactly 1 time\n",
      "    10303 (18.37%) aligned >1 times\n",
      "75.94% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "107867 reads; of these:\n",
      "  107867 (100.00%) were unpaired; of these:\n",
      "    55457 (51.41%) aligned 0 times\n",
      "    37253 (34.54%) aligned exactly 1 time\n",
      "    15157 (14.05%) aligned >1 times\n",
      "48.59% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "46847 reads; of these:\n",
      "  46847 (100.00%) were unpaired; of these:\n",
      "    6446 (13.76%) aligned 0 times\n",
      "    25875 (55.23%) aligned exactly 1 time\n",
      "    14526 (31.01%) aligned >1 times\n",
      "86.24% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "51760 reads; of these:\n",
      "  51760 (100.00%) were unpaired; of these:\n",
      "    2043 (3.95%) aligned 0 times\n",
      "    32766 (63.30%) aligned exactly 1 time\n",
      "    16951 (32.75%) aligned >1 times\n",
      "96.05% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "50448 reads; of these:\n",
      "  50448 (100.00%) were unpaired; of these:\n",
      "    2782 (5.51%) aligned 0 times\n",
      "    32949 (65.31%) aligned exactly 1 time\n",
      "    14717 (29.17%) aligned >1 times\n",
      "94.49% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "53566 reads; of these:\n",
      "  53566 (100.00%) were unpaired; of these:\n",
      "    37615 (70.22%) aligned 0 times\n",
      "    8896 (16.61%) aligned exactly 1 time\n",
      "    7055 (13.17%) aligned >1 times\n",
      "29.78% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "89684 reads; of these:\n",
      "  89684 (100.00%) were unpaired; of these:\n",
      "    13135 (14.65%) aligned 0 times\n",
      "    55645 (62.05%) aligned exactly 1 time\n",
      "    20904 (23.31%) aligned >1 times\n",
      "85.35% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n",
      "121582 reads; of these:\n",
      "  121582 (100.00%) were unpaired; of these:\n",
      "    5361 (4.41%) aligned 0 times\n",
      "    79500 (65.39%) aligned exactly 1 time\n",
      "    36721 (30.20%) aligned >1 times\n",
      "95.59% overall alignment rate\n",
      "[mpileup] 1 samples in 1 input files\n",
      "<mpileup> Set max per-file depth to 8000\n",
      "WARNING: package com.sun.java.swing.plaf.windows not in java.desktop\n",
      "WARNING: package sun.awt.windows not in java.desktop\n",
      "openjdk version \"11.0.1\" 2018-10-16 LTS\n",
      "OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)\n",
      "OpenJDK 64-Bit Server VM Zulu11.2+3 (build 11.0.1+13-LTS, mixed mode)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "chmod +x ./variants.sh\n",
    "./variants.sh NC_007898 /home/youngjb/Final_Project/Pipeline /home/youngjb/Final_Project"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  <font color='blue'> List All Generated Files </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BC_17.trim.bam      BC_21.trim.sam      BC_25.trim.bam      BC_29.trim.sam\r\n",
      "BC_17.trim.bam.bai  BC_21.trim.vcf      BC_25.trim.bam.bai  BC_29.trim.vcf\r\n",
      "BC_17.trim.fastq    BC_21.trim.vcf.idx  BC_25.trim.fastq    BC_29.trim.vcf.idx\r\n",
      "BC_17.trim.sam      BC_22.trim.bam      BC_25.trim.sam      BC_30.trim.bam\r\n",
      "BC_17.trim.vcf      BC_22.trim.bam.bai  BC_25.trim.vcf      BC_30.trim.bam.bai\r\n",
      "BC_17.trim.vcf.idx  BC_22.trim.fastq    BC_25.trim.vcf.idx  BC_30.trim.fastq\r\n",
      "BC_19.trim.bam      BC_22.trim.sam      BC_26.trim.bam      BC_30.trim.sam\r\n",
      "BC_19.trim.bam.bai  BC_22.trim.vcf      BC_26.trim.bam.bai  BC_30.trim.vcf\r\n",
      "BC_19.trim.fastq    BC_22.trim.vcf.idx  BC_26.trim.fastq    BC_30.trim.vcf.idx\r\n",
      "BC_19.trim.sam      BC_23.trim.bam      BC_26.trim.sam      NC_007898.1.bt2\r\n",
      "BC_19.trim.vcf      BC_23.trim.bam.bai  BC_26.trim.vcf      NC_007898.2.bt2\r\n",
      "BC_19.trim.vcf.idx  BC_23.trim.fastq    BC_26.trim.vcf.idx  NC_007898.3.bt2\r\n",
      "BC_20.trim.bam      BC_23.trim.sam      BC_28.trim.bam      NC_007898.4.bt2\r\n",
      "BC_20.trim.bam.bai  BC_23.trim.vcf      BC_28.trim.bam.bai  NC_007898.fasta\r\n",
      "BC_20.trim.fastq    BC_23.trim.vcf.idx  BC_28.trim.fastq    NC_007898.fasta.fai\r\n",
      "BC_20.trim.sam      BC_24.trim.bam      BC_28.trim.sam      NC_007898.gff\r\n",
      "BC_20.trim.vcf      BC_24.trim.bam.bai  BC_28.trim.vcf      NC_007898.rev.1.bt2\r\n",
      "BC_20.trim.vcf.idx  BC_24.trim.fastq    BC_28.trim.vcf.idx  NC_007898.rev.2.bt2\r\n",
      "BC_21.trim.bam      BC_24.trim.sam      BC_29.trim.bam\r\n",
      "BC_21.trim.bam.bai  BC_24.trim.vcf      BC_29.trim.bam.bai\r\n",
      "BC_21.trim.fastq    BC_24.trim.vcf.idx  BC_29.trim.fastq\r\n"
     ]
    }
   ],
   "source": [
    "ls ./Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  <font color='blue'> Integrative Genomics Viewer Extension Methods </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getFileTypesWithSuffix(directoryPath, suffix):\n",
    "    result = []\n",
    "    for entry in os.listdir(directoryPath):\n",
    "        if os.path.isfile(os.path.join(directoryPath, entry)):\n",
    "            if(entry.endswith(suffix)):\n",
    "                result.append(entry)\n",
    "    return result \n",
    "\n",
    "def addTrack(directoryPath, suffix, trackFormat, trackType, browser):\n",
    "    directory = directoryPath.split(\"/\")[len(directoryPath.split(\"/\"))-1]\n",
    "   \n",
    "    fileNames = getFileTypesWithSuffix(directoryPath, suffix)    \n",
    "    for fileName in fileNames:\n",
    "        browser.load_track({\"name\": fileName.split(\".\")[0], \"url\": os.path.join(\"files\", directory, fileName), \"format\": trackFormat, \"type\": trackType, \"indexed\": False })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  <font color='blue'> Load Integrative Genomics Viewer with Variant Track Data </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "igvBrowser = igv.Browser({\"reference\": {\n",
    "   \n",
    "    \"name\":  \"NC_007898\",\n",
    "    \"fastaURL\": \"files/Pipeline/NC_007898.fasta\",\n",
    "    \"indexed\": False}})\n",
    "\n",
    "addTrack(sequenceDirectoryFilePath, \".vcf\",  \"vcf\", \"variant\", igvBrowser)\n",
    "addTrack(sequenceDirectoryFilePath, \".gff\",  \"gff\", \"annotation\", igvBrowser)\n",
    "\n",
    "igvBrowser.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "rm -rf /mnt/c/Users/young/Documents/GitHub/data-science/Final_Project/*\n",
    "find /home/youngjb/Final_Project/ -maxdepth 1 -type f | xargs cp -t /mnt/c/Users/young/Documents/GitHub/data-science/Final_Project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
